tests/benchmarks/tpch/test_pyspark.py

```diff
@@ -0,0 +1,229 @@
+    return run_tpch_pyspark(tpch_pyspark_client, queries.q5)
+
+
+def test_query_6(tpch_pyspark_client):
+    return run_tpch_pyspark(tpch_pyspark_client, queries.q6)
+
+
+def test_query_7(tpch_pyspark_client):
+    return run_tpch_pyspark(tpch_pyspark_client, queries.q7)
+
+
+def make_pyspark_submit_args(spark_master):
+    return f"""
+        --packages {','.join(PACKAGES)}
+        --exclude-packages com.google.guava:guava
+        --master=spark://{spark_master}
+        pyspark-shell
+    """
+
+
+def fix_timestamp_ns_columns(query):
+    """
+    scale100 stores l_shipdate/o_orderdate as timestamp[us]
+    scale1000 stores l_shipdate/o_orderdate as timestamp[ns] which gives:
+        Illegal Parquet type: INT64 (TIMESTAMP(NANOS,true))
+    so we set spark.sql.legacy.parquet.nanosAsLong and then convert to timestamp.
+    """
+    for name in ("l_shipdate", "o_orderdate"):
+        query = re.sub(rf"\b{name}\b", f"to_timestamp(cast({name} as string))", query)
+    return query
+
+
+def run_tpch_pyspark(tpch_pyspark_client, module):
+    from .pyspark_queries.utils import get_or_create_spark
+
+    is_local = tpch_pyspark_client is None
+
+    async def _run_tpch(dask_scheduler):
+        # Connecting to the Spark Connect server doens't appear to work very well
+        # If we just submit this stuff to the scheduler and generate a SparkSession there
+        # that connects to the running master we can at least get some code running
+
+        if not is_local:
+            queries.utils.MASTER = f"spark://{get_ip()}:7077"
+        else:
+            queries.utils.MASTER = "local[*]"
+
+        os.environ["SPARK_HOME"] = str(pathlib.Path(pyspark.__file__).absolute().parent)
+        os.environ["PYSPARK_PYTHON"] = sys.executable
+        os.environ["PYSPARK_SUBMIT_ARGS"] = make_pyspark_submit_args(
+            queries.utils.MASTER
+        )
+
+        def _():
+            spark = get_or_create_spark(f"query{module.__name__}")
+
+            # scale1000 stored as timestamp[ns] which spark parquet
+            # can't use natively.
+            if ENABLED_DATASET == "scale 1000":
+                module.query = fix_timestamp_ns_columns(module.query)
+
+            module.setup(spark)  # read spark tables query will select from
+            if hasattr(module, "ddl"):
+                spark.sql(module.ddl)  # might create temp view
+            q_final = spark.sql(module.query)  # benchmark query
+            try:
+                # trigger materialization of df
+                return q_final.toJSON().collect()
+            finally:
+                spark.catalog.clearCache()
+                spark.sparkContext.stop()
+                spark.stop()
+
+        return await asyncio.to_thread(_)
+
+    if not is_local:
+        rows = tpch_pyspark_client.run_on_scheduler(_run_tpch)
+    else:
+        rows = asyncio.run(_run_tpch(None))  # running locally
+    print(f"Received {len(rows)} rows")
+
+
+class SparkMaster(SchedulerPlugin):
```

Comment:
I would also suggest as a follow up to

- Move the plugins into a spark.py submodule or something like this. That makes it easier to import if one runs a couple of ad-hoc tests. At least until we have this in coiled
- I would refactor the `run_tpch_pyspark` function slightly to allow it being used as `run_spark_query(dask_client, query)` for similar reasons.

Reply:
I think that for pyspark we merge it as-is, fix up spark outside of this repo, and then make this file like the others, where the query is just defined in the test.

Reply:
Future work regardless though.